The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/26330099427/job/77514677006) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/77514677006) [(plain)](https://github.com/rust-lang/rust/commit/ef04da37096403350e0159feaefd6e0d77ea3d52/checks/77514677006/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

---- [codegen] tests/codegen-llvm/call-site-inline-attributes.rs stdout ----
------FileCheck stdout------------------------------

------FileCheck stderr------------------------------
/checkout/tests/codegen-llvm/call-site-inline-attributes.rs:35:10: error: CHECK: expected string not found in input
//CHECK: attributes #[[NOINLINE]] = { noinline }
         ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen-llvm/call-site-inline-attributes/call-site-inline-attributes.ll:40:82: note: scanning from here
 call void @_RNvCs250flzSSnHr_27call_site_inline_attributes15alwaysinline_fn() #6
                                                                                 ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen-llvm/call-site-inline-attributes/call-site-inline-attributes.ll:40:82: note: with "NOINLINE" equal to "4"
 call void @_RNvCs250flzSSnHr_27call_site_inline_attributes15alwaysinline_fn() #6
                                                                                 ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen-llvm/call-site-inline-attributes/call-site-inline-attributes.ll:52:1: note: possible intended match here
attributes #1 = { noinline nounwind "target-cpu"="generic" }
^

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen-llvm/call-site-inline-attributes/call-site-inline-attributes.ll
Check file: /checkout/tests/codegen-llvm/call-site-inline-attributes.rs

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            1: ; ModuleID = 'call_site_inline_attributes.183c3d5a7d11aaf3-cgu.0' 
            2: source_filename = "call_site_inline_attributes.183c3d5a7d11aaf3-cgu.0" 
            3: target datalayout = "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-i128:128-n32:64-S128-ni:1:10:20" 
            4: target triple = "wasm32-unknown-wasip1" 
            5:  
            6: ; Function Attrs: nounwind 
            7: define dso_local void @calls_something_noinline() unnamed_addr #0 { 
            8: start: 
            9: ; call call_site_inline_attributes::noinline_fn 
           10:  call void @_RNvCs250flzSSnHr_27call_site_inline_attributes11noinline_fn() #4 
           11:  ret void 
           12: } 
           13:  
           14: ; call_site_inline_attributes::noinline_fn 
           15: ; Function Attrs: noinline nounwind 
           16: define internal void @_RNvCs250flzSSnHr_27call_site_inline_attributes11noinline_fn() unnamed_addr #1 { 
           17: start: 
           18:  ret void 
           19: } 
           20:  
           21: ; Function Attrs: nounwind 
           22: define dso_local void @calls_something_inline() unnamed_addr #0 { 
           23: start: 
           24: ; call call_site_inline_attributes::inlinehint_fn 
           25:  call void @_RNvCs250flzSSnHr_27call_site_inline_attributes13inlinehint_fn() #5 
           26:  ret void 
           27: } 
           28:  
           29: ; call_site_inline_attributes::inlinehint_fn 
           30: ; Function Attrs: inlinehint nounwind 
           31: define internal void @_RNvCs250flzSSnHr_27call_site_inline_attributes13inlinehint_fn() unnamed_addr #2 { 
           32: start: 
           33:  ret void 
           34: } 
           35:  
           36: ; Function Attrs: nounwind 
           37: define dso_local void @calls_something_alwaysinline() unnamed_addr #0 { 
           38: start: 
           39: ; call call_site_inline_attributes::alwaysinline_fn 
           40:  call void @_RNvCs250flzSSnHr_27call_site_inline_attributes15alwaysinline_fn() #6 
check:35'0                                                                                      X error: no match found
check:35'1                                                                                        with "NOINLINE" equal to "4"
           41:  ret void 
check:35'0     ~~~~~~~~~~
           42: } 
check:35'0     ~~
           43:  
check:35'0     ~
           44: ; call_site_inline_attributes::alwaysinline_fn 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           45: ; Function Attrs: alwaysinline nounwind 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           46: define internal void @_RNvCs250flzSSnHr_27call_site_inline_attributes15alwaysinline_fn() unnamed_addr #3 { 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           47: start: 
check:35'0     ~~~~~~~
           48:  ret void 
check:35'0     ~~~~~~~~~~
           49: } 
check:35'0     ~~
           50:  
check:35'0     ~
           51: attributes #0 = { nounwind "target-cpu"="generic" } 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           52: attributes #1 = { noinline nounwind "target-cpu"="generic" } 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:35'2     ?                                                             possible intended match
           53: attributes #2 = { inlinehint nounwind "target-cpu"="generic" } 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           54: attributes #3 = { alwaysinline nounwind "target-cpu"="generic" } 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           55: attributes #4 = { noinline nounwind } 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           56: attributes #5 = { inlinehint nounwind } 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           57: attributes #6 = { alwaysinline nounwind } 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           58:  
check:35'0     ~
           59: !llvm.ident = !{!0} 
check:35'0     ~~~~~~~~~~~~~~~~~~~~
           60:  
check:35'0     ~
           61: !0 = !{!"rustc version 1.98.0-nightly (0404a60bc 2026-05-23)"} 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>>>>>>

------------------------------------------

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen-llvm/call-site-inline-attributes/call-site-inline-attributes.ll" "/checkout/tests/codegen-llvm/call-site-inline-attributes.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
stdout: none
--- stderr -------------------------------
/checkout/tests/codegen-llvm/call-site-inline-attributes.rs:35:10: error: CHECK: expected string not found in input
//CHECK: attributes #[[NOINLINE]] = { noinline }
         ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen-llvm/call-site-inline-attributes/call-site-inline-attributes.ll:40:82: note: scanning from here
 call void @_RNvCs250flzSSnHr_27call_site_inline_attributes15alwaysinline_fn() #6
                                                                                 ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen-llvm/call-site-inline-attributes/call-site-inline-attributes.ll:40:82: note: with "NOINLINE" equal to "4"
 call void @_RNvCs250flzSSnHr_27call_site_inline_attributes15alwaysinline_fn() #6
                                                                                 ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen-llvm/call-site-inline-attributes/call-site-inline-attributes.ll:52:1: note: possible intended match here
attributes #1 = { noinline nounwind "target-cpu"="generic" }
^

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen-llvm/call-site-inline-attributes/call-site-inline-attributes.ll
Check file: /checkout/tests/codegen-llvm/call-site-inline-attributes.rs

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            1: ; ModuleID = 'call_site_inline_attributes.183c3d5a7d11aaf3-cgu.0' 
            2: source_filename = "call_site_inline_attributes.183c3d5a7d11aaf3-cgu.0" 
            3: target datalayout = "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-i128:128-n32:64-S128-ni:1:10:20" 
            4: target triple = "wasm32-unknown-wasip1" 
            5:  
            6: ; Function Attrs: nounwind 
            7: define dso_local void @calls_something_noinline() unnamed_addr #0 { 
            8: start: 
            9: ; call call_site_inline_attributes::noinline_fn 
           10:  call void @_RNvCs250flzSSnHr_27call_site_inline_attributes11noinline_fn() #4 
           11:  ret void 
           12: } 
           13:  
           14: ; call_site_inline_attributes::noinline_fn 
           15: ; Function Attrs: noinline nounwind 
           16: define internal void @_RNvCs250flzSSnHr_27call_site_inline_attributes11noinline_fn() unnamed_addr #1 { 
           17: start: 
           18:  ret void 
           19: } 
           20:  
           21: ; Function Attrs: nounwind 
           22: define dso_local void @calls_something_inline() unnamed_addr #0 { 
           23: start: 
           24: ; call call_site_inline_attributes::inlinehint_fn 
           25:  call void @_RNvCs250flzSSnHr_27call_site_inline_attributes13inlinehint_fn() #5 
           26:  ret void 
           27: } 
           28:  
           29: ; call_site_inline_attributes::inlinehint_fn 
           30: ; Function Attrs: inlinehint nounwind 
           31: define internal void @_RNvCs250flzSSnHr_27call_site_inline_attributes13inlinehint_fn() unnamed_addr #2 { 
           32: start: 
           33:  ret void 
           34: } 
           35:  
           36: ; Function Attrs: nounwind 
           37: define dso_local void @calls_something_alwaysinline() unnamed_addr #0 { 
           38: start: 
           39: ; call call_site_inline_attributes::alwaysinline_fn 
           40:  call void @_RNvCs250flzSSnHr_27call_site_inline_attributes15alwaysinline_fn() #6 
check:35'0                                                                                      X error: no match found
check:35'1                                                                                        with "NOINLINE" equal to "4"
           41:  ret void 
check:35'0     ~~~~~~~~~~
           42: } 
check:35'0     ~~
           43:  
check:35'0     ~
           44: ; call_site_inline_attributes::alwaysinline_fn 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           45: ; Function Attrs: alwaysinline nounwind 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           46: define internal void @_RNvCs250flzSSnHr_27call_site_inline_attributes15alwaysinline_fn() unnamed_addr #3 { 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           47: start: 
check:35'0     ~~~~~~~
           48:  ret void 
check:35'0     ~~~~~~~~~~
           49: } 
check:35'0     ~~
           50:  
check:35'0     ~
           51: attributes #0 = { nounwind "target-cpu"="generic" } 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           52: attributes #1 = { noinline nounwind "target-cpu"="generic" } 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:35'2     ?                                                             possible intended match
           53: attributes #2 = { inlinehint nounwind "target-cpu"="generic" } 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           54: attributes #3 = { alwaysinline nounwind "target-cpu"="generic" } 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           55: attributes #4 = { noinline nounwind } 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           56: attributes #5 = { inlinehint nounwind } 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           57: attributes #6 = { alwaysinline nounwind } 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           58:  
check:35'0     ~
           59: !llvm.ident = !{!0} 
check:35'0     ~~~~~~~~~~~~~~~~~~~~
           60:  
check:35'0     ~
           61: !0 = !{!"rustc version 1.98.0-nightly (0404a60bc 2026-05-23)"} 
check:35'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>>>>>>
------------------------------------------

---- [codegen] tests/codegen-llvm/call-site-inline-attributes.rs stdout end ----

```

</details>

--- a/compiler/rustc_codegen_llvm/src/attributes.rs
+++ b/compiler/rustc_codegen_llvm/src/attributes.rs
@@ -5,7 +5,7 @@ use rustc_hir::find_attr;
 use rustc_middle::middle::codegen_fn_attrs::{
     CodegenFnAttrFlags, CodegenFnAttrs, PatchableFunctionEntry, SanitizerFnAttrs, TargetFeature,
 };
-use rustc_middle::ty::{self, TyCtxt};
+use rustc_middle::ty::{self, Instance, TyCtxt};
 use rustc_session::config::{BranchProtection, FunctionReturn, OptLevel, PAuthKey, PacRet};
 use rustc_span::sym;
 use rustc_symbol_mangling::mangle_internal_symbol;
@@ -42,22 +42,31 @@ pub(crate) fn remove_string_attr_from_llfn(llfn: &Value, name: &str) {
 
 /// Get LLVM attribute for the provided inline heuristic.
 #[inline]
-fn inline_attr<'ll>(
+pub(crate) fn inline_attr<'tcx, 'll>(
     cx: &SimpleCx<'ll>,
-    sess: &Session,
-    inline: InlineAttr,
+    tcx: TyCtxt<'tcx>,
+    instance: Instance<'tcx>,
+    codegen_fn_attrs: &CodegenFnAttrs,
 ) -> Option<&'ll Attribute> {
-    if !sess.opts.unstable_opts.inline_llvm {
+    if !tcx.sess.opts.unstable_opts.inline_llvm {
         // disable LLVM inlining
         return Some(AttributeKind::NoInline.create_attr(cx.llcx));
     }
+
+    // `optnone` requires `noinline`
+    let inline = match (codegen_fn_attrs.inline, &codegen_fn_attrs.optimize) {
+        (_, OptimizeAttr::DoNotOptimize) => InlineAttr::Never,
+        (InlineAttr::None, _) if instance.def.requires_inline(tcx) => InlineAttr::Hint,
+        (inline, _) => inline,
+    };
+
     match inline {
         InlineAttr::Hint => Some(AttributeKind::InlineHint.create_attr(cx.llcx)),
         InlineAttr::Always | InlineAttr::Force { .. } => {
             Some(AttributeKind::AlwaysInline.create_attr(cx.llcx))
         }
         InlineAttr::Never => {
-            if sess.target.arch != Arch::AmdGpu {
+            if tcx.sess.target.arch != Arch::AmdGpu {
                 Some(AttributeKind::NoInline.create_attr(cx.llcx))
             } else {
                 None
@@ -412,14 +421,7 @@ pub(crate) fn llfn_attrs_from_instance<'ll, 'tcx>(
     }
 
     if let Some(instance) = instance {
-        // `optnone` requires `noinline`
-        let inline = match (codegen_fn_attrs.inline, &codegen_fn_attrs.optimize) {
-            (_, OptimizeAttr::DoNotOptimize) => InlineAttr::Never,
-            (InlineAttr::None, _) if instance.def.requires_inline(tcx) => InlineAttr::Hint,
-            (inline, _) => inline,
-        };
-
-        to_add.extend(inline_attr(cx, sess, inline));
+        to_add.extend(inline_attr(cx, tcx, instance, codegen_fn_attrs));
     }
 
     if sess.must_emit_unwind_tables() {

--- a/compiler/rustc_codegen_llvm/src/builder.rs
+++ b/compiler/rustc_codegen_llvm/src/builder.rs
@@ -1419,6 +1419,21 @@ impl<'a, 'll, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
             )
         };
 
+        if let Some(callee_instance) = callee_instance {
+            // Attributes on the function definition being called
+            let callee_attrs = self.cx.tcx.codegen_fn_attrs(callee_instance.def_id());
+
+            if let Some(inlining_rule) =
+                attributes::inline_attr(&self.cx, self.cx.tcx, callee_instance, callee_attrs)
+            {
+                attributes::apply_to_callsite(
+                    call,
+                    llvm::AttributePlace::Function,
+                    &[inlining_rule],
+                );
+            }
+        }
+
         if let Some(fn_abi) = fn_abi {
             fn_abi.apply_attrs_callsite(self, call);
         }

--- a/tests/codegen-llvm/call-site-inline-attributes.rs
+++ b/tests/codegen-llvm/call-site-inline-attributes.rs
@@ -1,0 +1,40 @@
+//@ compile-flags: -O -Zinline-mir=no -Cno-prepopulate-passes -Zmerge-functions=disabled
+
+#![crate_type = "lib"]
+
+// This test checks that we add inlinehint for #[inline], noinline for #[inline(never)], and
+// alwaysinline for #[inline(always)] to call sites.
+
+#[unsafe(no_mangle)]
+fn calls_something_noinline() {
+    // CHECK-LABEL @calls_something_noinline
+    // CHECK: call void @{{.*}}noinline_fn() #[[NOINLINE:[0-9]+]]
+    noinline_fn();
+}
+
+#[inline(never)]
+fn noinline_fn() {}
+
+#[unsafe(no_mangle)]
+fn calls_something_inline() {
+    // CHECK-LABEL @calls_something_inlinehint
+    // CHECK: call void @{{.*}}inlinehint_fn() #[[INLINEHINT:[0-9]+]]
+    inlinehint_fn();
+}
+
+#[inline]
+fn inlinehint_fn() {}
+
+#[unsafe(no_mangle)]
+fn calls_something_alwaysinline() {
+    // CHECK-LABEL @calls_something_alwaysinline
+    // CHECK: call void @{{.*}}alwaysinline_fn() #[[ALWAYSINLINE:[0-9]+]]
+    alwaysinline_fn();
+}
+
+#[inline(always)]
+fn alwaysinline_fn() {}
+
+//CHECK: attributes #[[NOINLINE]] = {{{.*}} noinline {{.*}}}
+//CHECK: attributes #[[INLINEHINT]] = {{{.*}} inlinehint {{.*}}}
+//CHECK: attributes #[[ALWAYSINLINE]] = {{{.*}} alwaysinline {{.*}}}

